logic_tree_tools.py

```diff
@@ -0,0 +1,972 @@
+    source_logic_tree_branch_set_short_names = [
+        x.short_name for x in logic_tree_pair.source_logic_tree.branch_sets
+    ]
+
+    # Initialize the results dictionary with empty dictionaries for each branch set
+    for (
+        source_logic_tree_branch_set_short_name
+    ) in source_logic_tree_branch_set_short_names:
+        results[source_logic_tree_branch_set_short_name] = {}
+
+    # Iterate through each branch set in the source logic tree
+    for branch_set_index, branch_set in enumerate(
+        logic_tree_pair.source_logic_tree.branch_sets
+    ):
+
+        if branch_set.short_name == "SLAB":
+            results["SLAB"] = {
+                "slab_only_branch": [copy.deepcopy(branch_set.branches[0])]
+            }
+
+        num_params = len(source_logic_tree_branch_params[branch_set.short_name])
+
+        # Iterate through each parameter in the branch set
+        for branch_param_idx in range(num_params):
+
+            selected_branches_per_param = []
+
+            # Allow both values at index branch_param_idx and keep everything else as the highest weighted branch
+            possible_values_for_this_param_index = source_logic_tree_branch_params[
+                branch_set.short_name
+            ][branch_param_idx]
+
+            if len(possible_values_for_this_param_index) == 1:
+                continue
+
+            hwb_values = (
+                source_logic_tree_highest_weighted_branches.branch_sets[
+                    branch_set_index
+                ]
+                .branches[0]
+                .values
+            )
+
+            # branch_values_to_find list will be modified for each param value
+            branch_values_to_find = copy.deepcopy(hwb_values)
+
+            for possible_param_value in possible_values_for_this_param_index:
+
+                branch_values_to_find[branch_param_idx] = possible_param_value
+
+                # This is needed as some string conversions return the text of interest surrounded by ' '
+                search_str = remove_single_quotes(str(branch_values_to_find))
+
+                for branch in branch_set.branches:
+
+                    if search_str == str(branch.values):
+                        selected_branches_per_param.append(copy.deepcopy(branch))
+
+            # The branches in selected_branches_per_param will be the only branches in this branch_set so
+            # their total weight needs to sum to 1.0
+
+            total_weight_selected_branches = sum(
+                [branch.weight for branch in selected_branches_per_param]
+            )
+            needed_scaling_factor = 1.0 / total_weight_selected_branches
+
+            for branch in selected_branches_per_param:
+                branch.weight *= needed_scaling_factor
+
+            assert sum(
+                [branch.weight for branch in selected_branches_per_param]
+            ), "Scaled weights of branches do not sum to 1.0"
```

Comment:
Is this not what the validate logic tree function checks? If so use that instead. Otherwise, change this to an if + valueerror

Reply:
The validated logic tree function is very similar but it's not exactly the same. The function's input is a whole logic but here we just have a collection of branches so it's probably simpler to leave it like this